packages/node-core/src/indexer/dictionary.service.ts

```diff
@@ -345,11 +342,19 @@ export class DictionaryService {
     dataSources: BlockHeightMap<DS[]>,
     buildDictionaryQueryEntries: (dataSources: DS[]) => DictionaryQueryEntry[]
   ): void {
-    this.queriesMap = dataSources.map((dataSources) => {
-      const query = buildDictionaryQueryEntries(dataSources);
-      const maxDsEndBlock = maxEndBlockHeight(dataSources);
-      return {query, maxDsEndBlock};
+    this.queriesMap = dataSources.map(buildDictionaryQueryEntries);
+    const newQueriesMap = this.queriesMap?.getAll() || new Map();
+
+    dataSources.getAll().forEach((ds, height) => {
+      const endBlock = maxEndBlockHeight(ds);
+      const queryDetails = this.queriesMap?.getDetails(height);
+
+      if (!queryDetails?.endHeight || queryDetails.endHeight > endBlock) {
+        newQueriesMap.set(endBlock + 1, []);
+      }
     });
+
+    this.queriesMap = new BlockHeightMap(newQueriesMap);
```

Comment:
I think it would make sense to build to datasources blockheight map correctly, then we know it matches up everywhere